DataStructures/AATree/AATree.cs

```diff
@@ -0,0 +1,552 @@
+
+        /// <summary>
+        /// Add a single element to the tree.
+        /// </summary>
+        /// <param name="key">The element to add to the tree.</param>
+        public void Add(TKey key)
+        {
+            root = Add(key, root);
+            Count++;
+        }
+
+        /// <summary>
+        /// Add multiple elements to the tree.
+        /// </summary>
+        /// <param name="keys">The elements to add to the tree.</param>
+        public void AddRange(IEnumerable<TKey> keys)
+        {
+            foreach (TKey key in keys)
+            {
+                root = Add(key, root);
+                Count++;
+            }
+        }
+
+        /// <summary>
+        /// Remove a single element from the tree.
+        /// </summary>
+        /// <param name="key">Element to remove.</param>
+        /// <returns>true if the element was successfully removed, false otherwise.</returns>
+        public bool Remove(TKey key)
+        {
+            if (!Contains(key, root))
+            {
+                return false;
+            }
+
+            root = Remove(key, root);
+            Count--;
+            return true;
+        }
+
+        /// <summary>
+        /// Checks if the specified element is in the tree.
+        /// </summary>
+        /// <param name="key">The element to look for.</param>
+        /// <returns>true if the element is in the tree, false otherwise.</returns>
+        public bool Contains(TKey key)
+        {
+            return Contains(key, root);
+        }
+
+        /// <summary>
+        /// Gets the largest element in the tree. (ie. the element in the right most node).
+        /// </summary>
+        /// <returns>The largest element in the tree according to the stored comparer.</returns>
+        /// <exception cref="InvalidOperationException">Thrown if the tree is empty.</exception>
+        public TKey GetMax()
+        {
+            if (root == null)
+            {
+                throw new InvalidOperationException("Tree is empty!");
+            }
+
+            return GetMax(root).Key;
+        }
+
+        /// <summary>
+        /// Gets the smallest element in the tree. (ie. the element in the left most node).
+        /// </summary>
+        /// <returns>The smallest element in the tree according to the stored comparer.</returns>
+        /// <throws>InvalidOperationException if the tree is empty.</throws>
+        public TKey GetMin()
+        {
+            if (root == null)
+            {
+                throw new InvalidOperationException("Tree is empty!");
+            }
+
+            return GetMin(root).Key;
+        }
+
+        /// <summary>
+        /// Gets all the elements in the tree in in-order order.
+        /// </summary>
+        /// <returns>Sequence of elements in in-order order.</returns>
+        public IEnumerable<TKey> GetKeysInOrder()
+        {
+            var result = new List<TKey>();
+
+            Action<AATreeNode<TKey>?>? inOrderWalk = null;
```

Comment:
Looks funny :D >?>?

Reply:
I'd suggest using a local function here, IMO it will look better and be more robust (it can't be null)